kernel/aster-nix/src/ipc/sem_set.rs

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{collections::btree_map::BTreeMap, vec::Vec};
+use core::time::Duration;
+
+use id_alloc::IdAlloc;
+use ostd::sync::{Mutex, RwMutex, RwMutexReadGuard, RwMutexWriteGuard};
+use spin::Once;
+
+use super::{key_t, IpcPermission};
+use crate::{
+    ipc::sem::Semaphore,
+    prelude::*,
+    process::{Gid, Uid},
+    time::SystemTime,
+};
+
+// The following constant values are derived from the default values in Linux.
+
+/// Maximum number of semaphore sets.
+pub const SEMMNI: usize = 32000;
+/// Maximum number of semaphores per semaphore ID.
+pub const SEMMSL: usize = 32000;
+/// Maximum number of seaphores in all semaphore sets.
+pub const SEMMNS: usize = SEMMNI * SEMMSL;
+/// Maximum number of operations for semop.
+pub const SEMOPM: usize = 500;
+/// MAximum semaphore value.
+pub const SEMVMX: i32 = 32767;
+/// Maximum value that can be recored for semaphore adjustment (SEM_UNDO).
+pub const SEMAEM: i32 = SEMVMX;
+
+static ID_ALLOCATOR: Once<Mutex<IdAlloc>> = Once::new();
+
+/// Semaphore sets in system
+static SEMAPHORE_SETS: RwMutex<BTreeMap<key_t, SemaphoreSet>> = RwMutex::new(BTreeMap::new());
+
+#[derive(Debug)]
+pub struct SemaphoreSet {
+    /// Number of semaphores in the set
+    nsems: usize,
+    /// Semaphores
+    sems: Vec<Semaphore>,
+    /// Semaphore permission
+    permission: IpcPermission,
+    /// Inner
+    inner: Mutex<Inner>,
```

Comment:
I didn't find the `AtomicDuration` in Asterinas repository. I found an `AtomicDuration` crate but it uses Apache License 2.0. Should we implement `AtomicDuration` in the Asterinas repository? 

I can use AtomicU64 or AtomicU128 for temporary use.

Reply:
> Inner only contains two times and the times do not need to be protected by `Mutex`, especially considering `Mutex` will bring performance cost.

I found something interesting in performance cost. I have tested four scenarios in `lat_sem`:

1. Use `Mutex` and `SystemTime::now()` to get time.
2. Use `Mutex` and remove `SystemTime::now()`.
3. Use `AtomicU64` and `SystemTime::now()` to get time.
4. Use `AtomicU64` and remove `SystemTime::now()`.

This will affect the `update_otime` function which is frequently used in `semop` syscall. The results are 0.935, 0.836, 0.896, and 0.797 μs separately.

The results show that replacing `Mutex` with `AtomicU64` will contribute 0.039 μs improvement while removing `SystemTime::now` will contribute to nearly 0.1 μs improvement. Why does `SystemTime::now` have such a high latency?

Reply:
I examined [the source code of `set_otime` in the Linux kernel](https://github.com/torvalds/linux/blob/d74da846046aeec9333e802f5918bd3261fb5509/ipc/sem.c#L1005). The current time is obtained using `ktime_get_real_seconds`. As indicated in the documentation for [this method](https://elixir.bootlin.com/linux/v6.0.9/source/kernel/time/timekeeping.c#L1020), this time is based on `CLOCK_REALTIME` and only includes the _seconds_ part.

In the current implementation, the current time is obtained using `System::now`, which internally uses `CLOCK_MONOTONIC`, which is not appropriate. Instead, `RealTimeClock::get().read_time()` should be used to obtain the current time. Additionally, you may consider using `RealTimeCoarseClock`, which is designed for obtaining low-precision time. You can experiment with both clocks to determine which provides better performance.

Reply:
> I didn't find the AtomicDuration in Asterinas repository. I found a crate implement AtomicDuration but it uses Apache License 2.0. Should we implement AtomicDuration in the Asterinas repository?

> I can use AtomicU64 for temporary use.

Using `AtomicU64` should be enough, since linux kernel only stores the _seconds_ part.